main.cc

```diff
@@ -1520,6 +1513,12 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
                 cdc_generation_service.stop().get();
             });
 
+            // XXX: stop_raft has to happen before query_processor, migration_manager and cdc_generation_service
+            // are stopped, since some groups keep using these services until are stopped inside stop_raft.
+            auto stop_raft = defer_verbose_shutdown("Raft", [&raft_gr] {
```

Comment:
This approach is wrong. We can't just keep moving raft_gr stopping later and later because "some groups" depend on some services.

We must stop the groups which have the dependency.

So the order of init/stop should be:

- start raft_group_registry
- start service X
- start a group which depends on X
- ... running ...
- stop the group which depends on X
- stop service X
- stop raft group registry.

OR:
- start services X, Y, ...
- start raft group registry and all groups which depend on X, Y, ...
- ... running ...
- stop raft group registry and all the groups which depend on X, Y...
- stop services X, Y, ...

In general, stopping order should be the reverse of starting order.

cc @kostja 

I see it was broken already before:
```cpp
            // XXX: stop_raft has to happen before query_processor and migration_manager
            // is stopped, since some groups keep using the query
            // processor until are stopped inside stop_raft.
            auto stop_raft = defer_verbose_shutdown("Raft", [&raft_gr] {
                raft_gr.stop().get();
            });
```

maybe this is a good opportunity to fix it.

Reply:
This is where group 0 is being setup:
```cpp

            // Set up group0 service earlier since it is needed by group0 setup just below
            ss.local().set_group0(group0_service);

            // Setup group0 early in case the node is bootstrapped already and the group exists.
            // Need to do it before allowing incomming messaging service connections since
            // storage proxy's and migration manager's verbs may access group0.
            // This will also disable migration manager schema pulls if needed.
            group0_service.setup_group0_if_exist(sys_ks.local(), ss.local(), qp.local(), mm.local(), cdc_generation_service.local()).get();

            // It's essential to load fencing_version prior to starting the messaging service,
            // since incoming messages may require fencing.
            // It is also necessary that token_metadata be initialized by this point and
            // token_metadata::version to be up to date, since when assigning fence_version we
            // validate it cannot be greater than token_metadata::version.
            // The setup_group0_if_exist call above waits for the raft log for group0 to be applied,
            // which ensures this.
            ss.local().update_fence_version(sys_ks.local().get_topology_fence_version().get()).get();

            with_scheduling_group(maintenance_scheduling_group, [&] {
                return messaging.invoke_on_all(&netw::messaging_service::start_listen, std::ref(token_metadata));
            }).get();

            with_scheduling_group(maintenance_scheduling_group, [&] {
                return ss.local().join_cluster(cdc_generation_service.local(), sys_dist_ks, proxy, qp.local());
            }).get();
```

Therefore this is where the dependency on CDC gen service and other services is introduced during startup.
Therefore, this is also where we should have a `stop_group0` deferred callback.

Reply:
This is also sloppy (the comment):
```cpp
            // group0 client exists only on shard 0.
            // The client has to be created before `stop_raft` since during
            // destruction it has to exist until raft_gr.stop() completes.
            service::raft_group0_client group0_client{raft_gr.local(), sys_ks.local()};
```

the comment argues that we need to stop_raft() before destroying group0_client.
But again, this is only because of the broken deinit which is missing a `stop_group0_client` step.

We must start fixing our service start/stop order, which should now be easy thanks to the sp/qp/sys_ks refactors. We should not be introducing more hacks into it or extending existing hacks.

cc @gleb-cloudius

Reply:
Every service construction+start should have a corresponding stop+destruct in a deferred callback that is deferred right after we start the service.

Similarly, if a cyclic dependency is introduced between services after they are started, there should be a deferred callback (right below introducing the dependency) that breaks the dependency.

So the code should look like:
```
X.start();
defer(stop_X);

Y.start(ref(X));
defer(stop_Y);

Z.start(ref(Y));
defer(stop_Z);

add_dependency_y_to_z(ref(Y), ref(Z));
defer(remove_dependency_y_to_z);
```

and so on.

cc @xemul

Reply:
> This is also sloppy (the comment):
> 
> ```c++
>             // group0 client exists only on shard 0.
>             // The client has to be created before `stop_raft` since during
>             // destruction it has to exist until raft_gr.stop() completes.
>             service::raft_group0_client group0_client{raft_gr.local(), sys_ks.local()};
> ```
> 
> the comment argues that we need to stop_raft() before destroying group0_client. But again, this is only because of the broken deinit which is missing a `stop_group0_client` step.

In general I agree that stopping all groups together  will not fly in the long run, but since the only group we have right now is group0 what you propose will simply rename things and will not change anything substantial. So here the comment will say that "The client has to be created before `stop_group0` since during destruction it has to exist until group0.stop() completes". Yes it is more correct. but today stopping group0 and stopping raft_gr is one and the same.

If you think this is something that should be done right now, I will not stop you.